### ISSUE TITLE
slidingWindow + key and function

### ISSUE THREAD

**Issue by ouhouhsami** [baconjs/bacon.js#570]
Hi, 

(Bacon.js, and FRP are awesome)

TL;DR

Is there a way to have slidingWindow to transform my EventStream from 
{time:1.2, frame:[1, 3, 5, 6, 5, 3, 4, 7, 2, 7, 8, 9, 10, 2, 4], metadata:{foo:bar}} 
to
{time:1.2, frame:[1, 3, 5], metadata:{foo:bar}} 
{time:2.2, frame:[6, 5, 3], metadata:{foo:bar}} 
{time:3.2, frame:[4, 7, 2], metadata:{foo:bar}} 
{time:4.2, frame:[7, 8, 9], metadata:{foo:bar}} 
{time:5.2, frame:[10, 2, 4], metadata:{foo:bar}} 

where, it would 'slide' on a key (here 'frame'), and modify, accordingly to a function another key (here 'time') ?

Long story

I use Bacons.js here: https://github.com/ouhouhsami/lfo-express. The idea is to transform getUserMedia as an Eventstream, and then, (audio)process it.

The raw experimentation works well to process the audio stream (in my case, I just want to process it, I don't want to ear the results).

Using slidingWindow on the EventStream, I would be able to make very precise FFT (Fast Fourier Transform), and other intensive computing algorithm.

My main concern is that I need to attach to each EventStream 2 other information: time and metadata. In the current use case, the time come from e.playbackTime (https://github.com/ouhouhsami/lfo-express/blob/master/baconGetUserMediaEventStream.js#L42). For the main EventStream I can have this kind of object: {time:<Float>, frame:<Float32Array>, metadata: <Object>}.
If, on this object I use map, I can easily use an audio processing function that returns  {time:<Float>, frame:<my_new_value>, metadata: <Object>}. 
But, I'd like to use slidingWindow on an object to get something like this: 
{time:1.2, frame:[1, 3, 5, 6, 5, 3, 4, 7, 2, 7, 8, 9, 10, 2, 4], metadata:{foo:bar}} 
to
{time:1.2, frame:[1, 3, 5], metadata:{foo:bar}} 
{time:2.2, frame:[6, 5, 3], metadata:{foo:bar}} 
{time:3.2, frame:[4, 7, 2], metadata:{foo:bar}} 
{time:4.2, frame:[7, 8, 9], metadata:{foo:bar}} 
{time:5.2, frame:[10, 2, 4], metadata:{foo:bar}} 

Thanks

**Comment by phadej** [comment]
I don't fully understand you.

Simple example:

```js
var stream = new Bacon.Bus(); // to show inputs
var output = stream.flatMap(function (val) {
  return Bacon.fromArray([val.slice(0, 2), val.slice(2)]); // (*)
});
output.log("out:");

stream.push([1,2,3,4,5]);
stream.push([6,7,8,9,0]);
```
produces:
```
out: [1, 2]
out: [3, 4, 5]
out: [6, 7]
out: [8, 9, 0]
```

---

So on line `(*)` we do pure data transformation, nothing Bacon specific. In your case you need to
```js
var val = {time:1.2, frame:[1, 3, 5, 6, 5, 3, 4, 7, 2, 7, 8, 9, 10, 2, 4], metadata:{foo:bar}};
```
into
```js
var res = [
  {time:1.2, frame:[1, 3, 5], metadata:{foo:bar}} ,
  {time:2.2, frame:[6, 5, 3], metadata:{foo:bar}},
  {time:3.2, frame:[4, 7, 2], metadata:{foo:bar}},
  {time:4.2, frame:[7, 8, 9], metadata:{foo:bar}},
  {time:5.2, frame:[10, 2, 4], metadata:{foo:bar}}
];
```
if I understood you right; Bacon itself doesn't help there at all.

**Comment by ouhouhsami** [comment]
Thanks for your answer.
But, my example was confusing, here is a complete one: 

From this EventStream
1 Event : {time:1.2, frame:[1, 3, 5, 6, 5, 3, 4, 7, 2, 7, 8, 9, 10, 2, 4], metadata:{foo:bar}}
2 Event : {time:9.2, frame:[5, 3, 6, 7, 8, 3, 4, 2, 6, 8, 9, 2, 11, 1, 1], metadata:{foo:baz}} 

to get
{time:1.2, frame:[1, 3, 5], metadata:{foo:bar}} 
{time:2.2, frame:[3, 5, 6], metadata:{foo:bar}} 
{time:3.2, frame:[5, 6, 5], metadata:{foo:bar}} 
{time:4.2, frame:[5, 3, 4], metadata:{foo:bar}} 
{time:5.2, frame:[4, 7, 2], metadata:{foo:bar}}
{time:6.2, frame:[2, 7, 8], metadata:{foo:bar}}
{time:7.2, frame:[8, 9, 10], metadata:{foo:bar}}
{time:8.2, frame:[10, 2, 4], metadata:{foo:bar}}
{time:8.2, frame:[2, 4, 5], metadata:{foo:bar}}
{time:9.2, frame:[5, 3, 6], metadata:{foo:baz}}
{time:10.2, frame:[3, 6, 7], metadata:{foo:baz}}
{time:11.2, frame:[7, 8, 3], metadata:{foo:baz}}
{time:12.2, frame:[3, 4, 2], metadata:{foo:baz}}
...

Best

**Comment by phadej** [comment]
Please, use formatting.

Transform step by step:
Split "compound" events into single events, and then use `slidingWindow`.
```js
{data: [1,2,3], meta: "first"}
{data: [4,5,6], meta: "second"}
```
to
```js
{data: 1, meta: "first"}
{data: 2, meta: "first"}
{data: 3, meta: "first"}
{data: 4, meta: "second"}
{data: 5, meta: "second"}
{data: 6, meta: "second"}
```
to
```js
[{data: 1, meta: "first"}, {data: 2, meta: "first"}, {data: 3, meta: "first"}]
[{data: 2, meta: "first"},{data: 3, meta: "first"},{data: 4, meta: "second"}]
...
```
to
```js
{data: [1,2,3], meta: "first"}
{data: [2,3,4], meta: "first"}
{data: [3,4,5], meta: "second"}
...
```
